Не могу разобраться в ошибке error C2668: 'TVector::TVector': ambiguous call to overloaded function. Появляется на стадии компиляции или code analysis из test_tmatrix, ссылаясь в самый конец файла.

--- a/include/utmatrix.h
+++ b/include/utmatrix.h
@@ -32,7 +32,6 @@ public:
 			else return b.Size;
   }
   */
-  TVector() {};
   TVector(int s = 10, int si = 0); 
   TVector(const TVector &v);                // конструктор копирования
   ~TVector();
@@ -71,19 +70,13 @@ public:
 template <class ValType>
 TVector<ValType>::TVector(int s, int si)
 {
-
-	if (s < 1 || s > MAX_VECTOR_SIZE)
+	if ((s < 1) || (s > MAX_VECTOR_SIZE))
 		throw s;
-	if (si < 0 || si > MAX_VECTOR_SIZE)
+	if ((si < 0) || (si >= MAX_VECTOR_SIZE))
 		throw si;
-	pVector = new ValType[Size];
 	Size = s;
 	StartIndex = si;
-	if (pVector != 0)
-		for (register int i = 0; i < Size; i++)
-			pVector[i] = 0;
-
-
+	pVector = new ValType[Size];
 } /*-------------------------------------------------------------------------*/
 
 template <class ValType> //конструктор копирования
@@ -91,8 +84,8 @@ TVector<ValType>::TVector(const TVector<ValType> &v)
 {
 	Size = v.Size;
 	StartIndex = v.StartIndex;
-	pVector = new ValType[v.Size];
-	for(register int i=0;i<v.Size;i++)
+	pVector = new ValType[Size];
+	for(register int i=0;i<Size;i++)
 		pVector[i]=v.pVector[i];
 } /*-------------------------------------------------------------------------*/
 
@@ -116,21 +109,18 @@ bool TVector<ValType>::operator==(const TVector &v) const
 {
 	if (v.Size != Size)  return false;
 	if (v.StartIndex != StartIndex) return false;
-	else
-	{
 		for (register int i = 0; i < Size; i++)
 		{
-			if v.pVector[i] != v.pVector[i] return false;
+			if (pVector[i] != v.pVector[i]) return false;
 		}
-	}
 	return true;
 } /*-------------------------------------------------------------------------*/
 
 template <class ValType> // сравнение
 bool TVector<ValType>::operator!=(const TVector &v) const
 {
-	if v == *this
-		return false 
+	if (v == *this)
+		return false;
 	else 
 		return true;
 } /*-------------------------------------------------------------------------*/
@@ -138,13 +128,16 @@ bool TVector<ValType>::operator!=(const TVector &v) const
 template <class ValType> // присваивание
 TVector<ValType>& TVector<ValType>::operator=(const TVector &v)
 {
-	if ((Size != v.Size) && (StartIndex!=v.StartIndex))
+	if (*this != v)
 	{
-		Size = v.Size;
-		StartIndex = v.StartIndex;
-		delete[] pVector;
-		pVector = 0;
-		pVector = new ValType[Size];
+		if ((Size != v.Size) && (StartIndex != v.StartIndex))
+		{
+			Size = v.Size;
+			StartIndex = v.StartIndex;
+			delete[] pVector;
+			pVector = 0;
+			pVector = new ValType[Size];
+		}
 	}
 	for (register int i = 0; i < Size; i++)
 		pVector[i] = v.pVector[i];

--- a/include/utmatrix.h
+++ b/include/utmatrix.h
@@ -70,7 +70,7 @@ public:
 template <class ValType>
 TVector<ValType>::TVector(int s, int si)
 {
-	if ((s < 1) || (s > MAX_VECTOR_SIZE))
+	if ((s < 1) || (s >= MAX_VECTOR_SIZE))
 		throw s;
 	if ((si < 0) || (si >= MAX_VECTOR_SIZE))
 		throw si;
@@ -99,9 +99,8 @@ TVector<ValType>::~TVector()
 template <class ValType> // доступ
 ValType& TVector<ValType>::operator[](int pos)
 {
-	if (pos<0 || pos>Size)
-		throw pos;
-	return pVector[abs(pos - StartIndex)]; //if pos<SI
+	if ((pos - StartIndex >= 0) && (pos - StartIndex < Size))
+		return pVector[pos - StartIndex]; else throw(pos);
 } /*-------------------------------------------------------------------------*/
 
 template <class ValType> // сравнение
@@ -132,12 +131,12 @@ TVector<ValType>& TVector<ValType>::operator=(const TVector &v)
 	{
 		if (Size != v.Size)
 		{
-			Size = v.Size;
-			StartIndex = v.StartIndex;
 			delete[] pVector;
 			pVector = 0;
-			pVector = new ValType[Size];
+			pVector = new ValType[v.Size];
 		}
+		Size = v.Size;
+		StartIndex = v.StartIndex;
 	for (register int i = 0; i < Size; i++)
 		pVector[i] = v.pVector[i];
 	}
@@ -237,7 +236,7 @@ template <class ValType>
 TMatrix<ValType>::TMatrix(int s) : TVector<TVector<ValType> >(s)
 {
 	//Size = s;
-	if (s<1 || s>MAX_MATRIX_SIZE) throw s;
+	if (s>MAX_MATRIX_SIZE) throw s;
 	for (register int i = 0; i < s; i++)
 	{
 		pVector[i] = TVector<ValType>(s - i, i);
@@ -273,21 +272,16 @@ bool TMatrix<ValType>::operator!=(const TMatrix<ValType> &mt) const
 template <class ValType> // присваивание
 TMatrix<ValType>& TMatrix<ValType>::operator=(const TMatrix<ValType> &mt)
 {
-	StartIndex = mt.StartIndex;
+	
 	if (Size != mt.Size)
 	{
 		delete[] pVector;
-		Size = mt.Size;
-		pVector = new TVector<ValType>[Size];
-		for (register int i = 0; i < Size; i++)
-		{
-			pVector[i] = mt.pVector[i];
-		}
+		pVector = new TVector<ValType>[mt.Size];
 	}
+	Size = mt.Size;
+	StartIndex = mt.StartIndex;
 	for (register int i = 0; i < Size; i++)
-	{
 		pVector[i] = mt.pVector[i];
-	}
 	return *this;
 } /*-------------------------------------------------------------------------*/
 

--- a/include/utmatrix.h
+++ b/include/utmatrix.h
@@ -20,51 +20,51 @@ template <class ValType>
 class TVector
 {
 protected:
-  ValType *pVector;
-  int Size;       // размер вектора
-  int StartIndex; // индекс первого элемента вектора
+	ValType *pVector;
+	int Size;       // размер вектора
+	int StartIndex; // индекс первого элемента вектора
 public:
 	/*
-  int MaxSize(const TVector &a, const TVector &b)
-  {
-			if (a.Size == b.Size) return a.Size;
-			if (a.Size > b.Size) return a.Size
-			else return b.Size;
-  }
-  */
-  TVector(int s = 10, int si = 0); 
-  TVector(const TVector &v);                // конструктор копирования
-  ~TVector();
-  int GetSize()      { return Size;       } // размер вектора
-  int GetStartIndex(){ return StartIndex; } // индекс первого элемента
-  ValType& operator[](int pos);             // доступ
-  bool operator==(const TVector &v) const;  // сравнение
-  bool operator!=(const TVector &v) const;  // сравнение
-  TVector& operator=(const TVector &v);     // присваивание
+	int MaxSize(const TVector &a, const TVector &b)
+	{
+	if (a.Size == b.Size) return a.Size;
+	if (a.Size > b.Size) return a.Size
+	else return b.Size;
+	}
+	*/
+	TVector(int s = 10, int si = 0);
+	TVector(const TVector &v);                // конструктор копирования
+	~TVector();
+	int GetSize() { return Size; } // размер вектора
+	int GetStartIndex() { return StartIndex; } // индекс первого элемента
+	ValType& operator[](int pos);             // доступ
+	bool operator==(const TVector &v) const;  // сравнение
+	bool operator!=(const TVector &v) const;  // сравнение
+	TVector& operator=(const TVector &v);     // присваивание
 
-  // скалярные операции
-  TVector  operator+(const ValType &val);   // прибавить скаляр
-  TVector  operator-(const ValType &val);   // вычесть скаляр
-  TVector  operator*(const ValType &val);   // умножить на скаляр
+											  // скалярные операции
+	TVector  operator+(const ValType &val);   // прибавить скаляр
+	TVector  operator-(const ValType &val);   // вычесть скаляр
+	TVector  operator*(const ValType &val);   // умножить на скаляр
 
-  // векторные операции
-  TVector  operator+(const TVector &v);     // сложение
-  TVector  operator-(const TVector &v);     // вычитание
-  ValType  operator*(const TVector &v);     // скалярное произведение
+											  // векторные операции
+	TVector  operator+(const TVector &v);     // сложение
+	TVector  operator-(const TVector &v);     // вычитание
+	ValType  operator*(const TVector &v);     // скалярное произведение
 
-  // ввод-вывод
-  friend istream& operator>>(istream &in, TVector &v)
-  {
-    for (int i = 0; i < v.Size; i++)
-      in >> v.pVector[i];
-    return in;
-  }
-  friend ostream& operator<<(ostream &out, const TVector &v)
-  {
-    for (int i = 0; i < v.Size; i++)
-      out << v.pVector[i] << ' ';
-    return out;
-  }
+											  // ввод-вывод
+	friend istream& operator >> (istream &in, TVector &v)
+	{
+		for (int i = 0; i < v.Size; i++)
+			in >> v.pVector[i];
+		return in;
+	}
+	friend ostream& operator<<(ostream &out, const TVector &v)
+	{
+		for (int i = 0; i < v.Size; i++)
+			out << v.pVector[i] << ' ';
+		return out;
+	}
 };
 
 template <class ValType>
@@ -85,8 +85,8 @@ TVector<ValType>::TVector(const TVector<ValType> &v)
 	Size = v.Size;
 	StartIndex = v.StartIndex;
 	pVector = new ValType[Size];
-	for(register int i=0;i<Size;i++)
-		pVector[i]=v.pVector[i];
+	for (register int i = 0; i<Size; i++)
+		pVector[i] = v.pVector[i];
 } /*-------------------------------------------------------------------------*/
 
 template <class ValType>
@@ -99,9 +99,9 @@ TVector<ValType>::~TVector()
 template <class ValType> // доступ
 ValType& TVector<ValType>::operator[](int pos)
 {
-	if (pos<0 || pos>MAX_VECTOR_SIZE)
+	if (pos<0 || pos>Size)
 		throw pos;
-	return pVector[abs(pos-StartIndex)]; //if pos<SI
+	return pVector[abs(pos - StartIndex)]; //if pos<SI
 } /*-------------------------------------------------------------------------*/
 
 template <class ValType> // сравнение
@@ -109,10 +109,10 @@ bool TVector<ValType>::operator==(const TVector &v) const
 {
 	if (v.Size != Size)  return false;
 	if (v.StartIndex != StartIndex) return false;
-		for (register int i = 0; i < Size; i++)
-		{
-			if (pVector[i] != v.pVector[i]) return false;
-		}
+	for (register int i = 0; i < Size; i++)
+	{
+		if (pVector[i] != v.pVector[i]) return false;
+	}
 	return true;
 } /*-------------------------------------------------------------------------*/
 
@@ -121,7 +121,7 @@ bool TVector<ValType>::operator!=(const TVector &v) const
 {
 	if (v == *this)
 		return false;
-	else 
+	else
 		return true;
 } /*-------------------------------------------------------------------------*/
 
@@ -130,7 +130,7 @@ TVector<ValType>& TVector<ValType>::operator=(const TVector &v)
 {
 	if (*this != v)
 	{
-		if ((Size != v.Size) && (StartIndex != v.StartIndex))
+		if (Size != v.Size)
 		{
 			Size = v.Size;
 			StartIndex = v.StartIndex;
@@ -138,11 +138,11 @@ TVector<ValType>& TVector<ValType>::operator=(const TVector &v)
 			pVector = 0;
 			pVector = new ValType[Size];
 		}
-	}
 	for (register int i = 0; i < Size; i++)
 		pVector[i] = v.pVector[i];
+	}
 	return *this;
-	
+
 } /*-------------------------------------------------------------------------*/
 
 template <class ValType> // прибавить скаляр
@@ -150,7 +150,7 @@ TVector<ValType> TVector<ValType>::operator+(const ValType &val)
 {
 	TVector tmp(Size, StartIndex);
 	for (register int i = 0; i < Size; i++)
-		tmp.pVector[i] = v.pVector[i] + val;
+		tmp.pVector[i] = pVector[i] + val;
 	return tmp;
 } /*-------------------------------------------------------------------------*/
 
@@ -159,7 +159,7 @@ TVector<ValType> TVector<ValType>::operator-(const ValType &val)
 {
 	TVector tmp(Size, StartIndex);
 	for (register int i = 0; i < Size; i++)
-		tmp.pVector[i] = v.pVector[i] - val;
+		tmp.pVector[i] = pVector[i] - val;
 	return tmp;
 } /*-------------------------------------------------------------------------*/
 
@@ -168,73 +168,73 @@ TVector<ValType> TVector<ValType>::operator*(const ValType &val)
 {
 	TVector tmp(Size, StartIndex);
 	for (register int i = 0; i < Size; i++)
-		tmp.pVector[i] = v.pVector[i] * val;
+		tmp.pVector[i] = pVector[i] * val;
 	return tmp;
 } /*-------------------------------------------------------------------------*/
 
 template <class ValType> // сложение
 TVector<ValType> TVector<ValType>::operator+(const TVector<ValType> &v)
 {
-	if ((Size != v.Size) && (v.StartIndex != StartIndex)) throw v;
-		TVector tmp(Size, StartIndex);
-		for (register int i = 0; i < Size; i++)
-			tmp.pVector[i] = pVector[i] + v.pVector[i];
-		return tmp;
-		
+	if (Size != v.Size) throw v;
+	TVector tmp(Size, StartIndex);
+	for (register int i = 0; i < Size; i++)
+		tmp.pVector[i] = pVector[i] + v.pVector[i];
+	return tmp;
+
 } /*-------------------------------------------------------------------------*/
 
 template <class ValType> // вычитание
 TVector<ValType> TVector<ValType>::operator-(const TVector<ValType> &v)
 {
-	if ((Size != v.Size) && (v.StartIndex != StartIndex)) throw v;
-		TVector tmp(Size, StartIndex);
-		for (register int i = 0; i < Size; i++)
-			tmp.pVector[i] = pVector[i] - v.pVector[i];
-		return tmp;
+	if (Size != v.Size) throw v;
+	TVector tmp(Size, StartIndex);
+	for (register int i = 0; i < Size; i++)
+		tmp.pVector[i] = pVector[i] - v.pVector[i];
+	return tmp;
 } /*-------------------------------------------------------------------------*/
 
 template <class ValType> // скалярное произведение
 ValType TVector<ValType>::operator*(const TVector<ValType> &v)
 {
-	if ((Size != v.Size) && (v.StartIndex != StartIndex)) throw v;
-		TVector tmp(Size, StartIndex);
-		for (register int i = 0; i < Size; i++)
-			tmp.pVector[i] = pVector[i] * v.pVector[i];
-		return tmp;
+	if (Size != v.Size) throw v;
+	ValType res=0;
+	for (register int i = 0; i < Size; i++)
+		res= res + (pVector[i] * v.pVector[i]);
+	return res;
 } /*-------------------------------------------------------------------------*/
 
 
-// Верхнетреугольная матрица
+  // Верхнетреугольная матрица
 template <class ValType>
 class TMatrix : public TVector<TVector<ValType> >
 {
 public:
-  TMatrix(int s = 10);           // изначально s=10;                 
-  TMatrix(const TMatrix &mt);                    // копирование
-  TMatrix(const TVector<TVector<ValType> > &mt); // преобразование типа
-  bool operator==(const TMatrix &mt) const;      // сравнение
-  bool operator!=(const TMatrix &mt) const;      // сравнение
-  TMatrix& operator= (const TMatrix &mt);        // присваивание
-  TMatrix  operator+ (const TMatrix &mt);        // сложение
-  TMatrix  operator- (const TMatrix &mt);        // вычитание
+	TMatrix(int s = 10);           // изначально s=10;                 
+	TMatrix(const TMatrix &mt);                    // копирование
+	TMatrix(const TVector<TVector<ValType> > &mt); // преобразование типа
+	bool operator==(const TMatrix &mt) const;      // сравнение
+	bool operator!=(const TMatrix &mt) const;      // сравнение
+	TMatrix& operator= (const TMatrix &mt);        // присваивание
+	TMatrix  operator+ (const TMatrix &mt);        // сложение
+	TMatrix  operator- (const TMatrix &mt);        // вычитание
 
-  // ввод / вывод
-  friend istream& operator>>(istream &in, TMatrix &mt)
-  {
-    for (int i = 0; i < mt.Size; i++)
-      in >> mt.pVector[i];
-    return in;
-  }
-  friend ostream & operator<<( ostream &out, const TMatrix &mt)
-  {
-    for (int i = 0; i < mt.Size; i++)
-      out << mt.pVector[i] << endl;
-    return out;
-  }
+												   // ввод / вывод
+	friend istream& operator >> (istream &in, TMatrix &mt)
+	{
+		for (int i = 0; i < mt.Size; i++)
+			in >> mt.pVector[i];
+		return in;
+	}
+	friend ostream & operator<<(ostream &out, const TMatrix &mt)
+	{
+		for (int i = 0; i < mt.Size; i++)
+			out << mt.pVector[i] << endl;
+		return out;
+	}
 };
 
 template <class ValType>
-TMatrix<ValType>::TMatrix(int s): TVector<TVector<ValType> >(s)
+TMatrix<ValType>::TMatrix(int s) : TVector<TVector<ValType> >(s)
 {
 	//Size = s;
 	if (s<1 || s>MAX_MATRIX_SIZE) throw s;
@@ -245,19 +245,18 @@ TMatrix<ValType>::TMatrix(int s): TVector<TVector<ValType> >(s)
 } /*-------------------------------------------------------------------------*/
 
 template <class ValType> // конструктор копирования
-TMatrix<ValType>::TMatrix(const TMatrix<ValType> &mt):
-  TVector<TVector<ValType> >(mt) {} //auto
+TMatrix<ValType>::TMatrix(const TMatrix<ValType> &mt) :
+	TVector<TVector<ValType> >(mt) {} //auto
 
 template <class ValType> // конструктор преобразования типа
-TMatrix<ValType>::TMatrix(const TVector<TVector<ValType> > &mt):
-  TVector<TVector<ValType> >(mt) {} //auto
+TMatrix<ValType>::TMatrix(const TVector<TVector<ValType> > &mt) :
+	TVector<TVector<ValType> >(mt) {} //auto
 
 template <class ValType> // сравнение
 bool TMatrix<ValType>::operator==(const TMatrix<ValType> &mt) const
 {
 	if (Size != mt.Size) return false;
-	if (StartIndex != mt.StartIndex) return false;
-	for (register int i = 0; i < si++)
+	for (register int i = 0; i < Size; i++)
 	{
 		if (pVector[i] != mt.pVector[i]) return false;
 	}
@@ -267,7 +266,7 @@ bool TMatrix<ValType>::operator==(const TMatrix<ValType> &mt) const
 template <class ValType> // сравнение
 bool TMatrix<ValType>::operator!=(const TMatrix<ValType> &mt) const
 {
-	if *this == mt return false
+	if (*this == mt) return false;
 	else return true;
 } /*-------------------------------------------------------------------------*/
 
@@ -280,10 +279,10 @@ TMatrix<ValType>& TMatrix<ValType>::operator=(const TMatrix<ValType> &mt)
 		delete[] pVector;
 		Size = mt.Size;
 		pVector = new TVector<ValType>[Size];
-			for (register int i = 0; i < Size; i++)
-			{
-				pVector[i] = mt.pVector[i];
-			}
+		for (register int i = 0; i < Size; i++)
+		{
+			pVector[i] = mt.pVector[i];
+		}
 	}
 	for (register int i = 0; i < Size; i++)
 	{
@@ -304,6 +303,6 @@ TMatrix<ValType> TMatrix<ValType>::operator-(const TMatrix<ValType> &mt)
 	return TVector<TVector<ValType> >::operator-(mt);
 } /*-------------------------------------------------------------------------*/
 
-// TVector О3 Л2 П4 С6
-// TMatrix О2 Л2 П3 С3
+  // TVector О3 Л2 П4 С6
+  // TMatrix О2 Л2 П3 С3
 #endif

